File: src/main/java/com/team3/gudit/goods/mapper/GoodsMapper.java
package com.team3.gudit.goods.mapper;

import com.team3.gudit.goods.domain.entity.Goods;
import com.team3.gudit.goods.domain.enums.GoodsStatus;
import com.team3.gudit.goods.dto.request.GoodsCreateRequest;
import com.team3.gudit.goods.dto.request.GoodsStatusUpdateRequest;
import com.team3.gudit.goods.dto.request.GoodsUpdateRequest;
import com.team3.gudit.goods.dto.response.*;
import org.springframework.stereotype.Component;

@Component
public class GoodsMapper {

    public Goods toEntity(GoodsCreateRequest request) {
        return toEntity(request, request.imageUrl());
    }

    public Goods toEntity(GoodsCreateRequest request, String imageUrl) {
        return Goods.builder()
                .name(request.name())
                .description(request.description())
                .price(request.price())
                .imageUrl(imageUrl)
                .status(GoodsStatus.ACTIVE)
                .build();
    }

    public void updateEntity(
            Goods goods,
            GoodsUpdateRequest request,
            String imageUrl
    ) {
        goods.updateGoodsInfo(
                request.name(),
                request.description(),
                request.price(),
                imageUrl
        );
    }

    public void updateStatus(Goods goods, GoodsStatusUpdateRequest request) {
        goods.updateGoodsStatus(request.status());
    }

    public GoodsCreateResponse toCreateResponse(Goods goods) {
        return GoodsCreateResponse.builder()
                .id(goods.getId())
                .name(goods.getName())
                .description(goods.getDescription())
                .price(goods.getPrice())
                .imageUrl(goods.getImageUrl())
                .status(goods.getStatus())
                .createdAt(goods.getCreatedAt())
                .build();
    }

    public GoodsListResponse toListResponse(Goods goods) {
        return GoodsListResponse.builder()
                .id(goods.getId())
                .name(goods.getName())
                .price(goods.getPrice())
                .imageUrl(goods.getImageUrl())
                .status(goods.getStatus())
                .createdAt(goods.getCreatedAt())
                .build();
    }

    public GoodsDetailResponse toDetailResponse(Goods goods) {
        return GoodsDetailResponse.builder()
                .goodsId(goods.getId())
                .name(goods.getName())
                .description(goods.getDescription())
                .price(goods.getPrice())
                .imageUrl(goods.getImageUrl())
                .status(goods.getStatus())
                .createdAt(goods.getCreatedAt())
                .build();
    }


    public GoodsUpdateResponse toUpdateResponse(Goods goods) {
        return GoodsUpdateResponse.builder()
                .id(goods.getId())
                .name(goods.getName())
                .description(goods.getDescription())
                .price(goods.getPrice())
                .imageUrl(goods.getImageUrl())
                .status(goods.getStatus())
                .build();
    }

    public GoodsStatusUpdateResponse toStatusUpdateResponse(Goods goods) {
        return GoodsStatusUpdateResponse.builder()
                .id(goods.getId())
                .status(goods.getStatus())
                .updatedAt(goods.getUpdatedAt())
                .build();
    }

}

File: src/main/kotlin/com/team3/gudit/goods/mapper/GoodsMapper.kt
package com.team3.gudit.goods.mapper

import com.team3.gudit.goods.domain.entity.Goods
import com.team3.gudit.goods.domain.enums.GoodsStatus
import com.team3.gudit.goods.dto.request.GoodsCreateRequest
import com.team3.gudit.goods.dto.request.GoodsStatusUpdateRequest
import com.team3.gudit.goods.dto.request.GoodsUpdateRequest
import com.team3.gudit.goods.dto.response.GoodsCreateResponse
import com.team3.gudit.goods.dto.response.GoodsDetailResponse
import com.team3.gudit.goods.dto.response.GoodsListResponse
import com.team3.gudit.goods.dto.response.GoodsStatusUpdateResponse
import com.team3.gudit.goods.dto.response.GoodsUpdateResponse
import org.springframework.stereotype.Component

@Component
class GoodsMapper {
    fun toEntity(request: GoodsCreateRequest): Goods = toEntity(request, request.imageUrl)

    fun toEntity(
        request: GoodsCreateRequest,
        imageUrl: String?,
    ): Goods =
        Goods.builder()
            .name(request.name)
            .description(request.description)
            .price(request.price)
            .imageUrl(imageUrl)
            .status(GoodsStatus.ACTIVE)
            .build()

    fun updateEntity(
        goods: Goods,
        request: GoodsUpdateRequest,
        imageUrl: String?,
    ) {
        goods.updateGoodsInfo(
            request.name,
            request.description,
            request.price,
            imageUrl,
        )
    }

    fun updateStatus(
        goods: Goods,
        request: GoodsStatusUpdateRequest,
    ) {
        goods.updateGoodsStatus(request.status)
    }

    fun toCreateResponse(goods: Goods): GoodsCreateResponse =
        GoodsCreateResponse.builder()
            .id(goods.id)
            .name(goods.name)
            .description(goods.description)
            .price(goods.price)
            .imageUrl(goods.imageUrl)
            .status(goods.status)
            .createdAt(goods.createdAt)
            .build()

    fun toListResponse(goods: Goods): GoodsListResponse =
        GoodsListResponse.builder()
            .id(goods.id)
            .name(goods.name)
            .price(goods.price)
            .imageUrl(goods.imageUrl)
            .status(goods.status)
            .createdAt(goods.createdAt)
            .build()

    fun toDetailResponse(goods: Goods): GoodsDetailResponse =
        GoodsDetailResponse.builder()
            .goodsId(goods.id)
            .name(goods.name)
            .description(goods.description)
            .price(goods.price)
            .imageUrl(goods.imageUrl)
            .status(goods.status)
            .createdAt(goods.createdAt)
            .build()

    fun toUpdateResponse(goods: Goods): GoodsUpdateResponse =
        GoodsUpdateResponse.builder()
            .id(goods.id)
            .name(goods.name)
            .description(goods.description)
            .price(goods.price)
            .imageUrl(goods.imageUrl)
            .status(goods.status)
            .build()

    fun toStatusUpdateResponse(goods: Goods): GoodsStatusUpdateResponse =
        GoodsStatusUpdateResponse.builder()
            .id(goods.id)
            .status(goods.status)
            .updatedAt(goods.updatedAt)
            .build()
}
